client/azure_client_set.go

```diff
@@ -160,7 +160,7 @@ func prepareClient(client *autorest.Client, authorizer autorest.Authorizer, part
 	return client
 }
 
-func newDeploymentsClient(authorizer autorest.Authorizer, subscriptionID, partnerID string) (*resources.DeploymentsClient, error) {
+func newDeploymentsClient(authorizer autorest.Authorizer, subscriptionID, partnerID string) (interface{}, error) {
```

Comment:
why not use `toDeploymentsClient` inside this function?

Reply:
It's so we can have more generic (yet simple) code for Azure client creation in client factory. It was initially Tuomas' suggestion.

For example this is now the creation/getting of `DeploymentsClient`:
```
func (f *Factory) GetDeploymentsClient(cr v1alpha1.AzureConfig) (*resources.DeploymentsClient, error) {
	client, err := f.getClient(cr, "DeploymentsClient", newDeploymentsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toDeploymentsClient(client), nil
}
```

And this the same for `GroupsClient`:
```
func (f *Factory) GetGroupsClient(cr v1alpha1.AzureConfig) (*resources.GroupsClient, error) {
	client, err := f.getClient(cr, "GroupsClient", newGroupsClient)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	return toGroupsClient(client), nil
}
```

So in both cases the client is created with generic function that is returning `interface{}`:
```
type clientCreatorFunc func(autorest.Authorizer, string, string) (interface{}, error)

func (f *Factory) getClient(cr v1alpha1.AzureConfig, clientType string, createClient clientCreatorFunc) (interface{}, error) {
// ...
}
```

Without returning `interface{}` in `newDeploymentsClient` (and other similar functions), we would have to do something like this for creating clients in the factory, having a wrapper function to return an `interface{}:
```
func (f *Factory) GetDeploymentsClient(cr v1alpha1.AzureConfig) (*resources.DeploymentsClient, error) {
	createClientFunc := func(authorizer autorest.Authorizer, subscriptionID string, partnerID string) (interface{}, error) {
		return newDeploymentsClient(authorizer, subscriptionID, partnerID)
	}
	client, err := f.getClient(cr, "DeploymentsClient", createClientFunc)
	if err != nil {
		return nil, microerror.Mask(err)
	}


	return client.(*resources.DeploymentsClient), nil
}
```

So the whole idea was to make new client factory code simpler and cleaner, and let the old code take a bit of the burden :) [Here is a commit with before and after this change](https://github.com/giantswarm/azure-operator/pull/853/commits/1223182eaae7acf8d5a2e7f6d0078ccfc8601230).

Reply:
I understand now. It's better like this.